### PR TITLE
Simple block allocator

### PR DESCRIPTION
**Pull request by wmitros**
See #3

### CODE REVIEW

File: src/fs/block_allocator.hh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include <unordered_map>
+#include <queue>
+
+namespace seastar {
+
+class block_allocator {
+    std::unordered_map<size_t, bool> _block_pool;
+    std::queue<size_t> _free_blocks;
+public:
+    block_allocator(std::unordered_map<size_t, bool>, std::queue<size_t>);
```
Comment (tropuq) [inline src/fs/block_allocator.hh]: What is the purpose of passing that unordered_map to the block_allocator constructor?

File: src/fs/block_allocator.hh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+#pragma once
+
+#include <unordered_map>
+#include <queue>
+
+namespace seastar {
+
+class block_allocator {
```
Comment (tropuq) [inline src/fs/block_allocator.hh]: What do you think about adding everything to the fs namespace?

File: src/fs/block_allocator.hh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    std::unordered_map<size_t, bool> _block_pool;
```
Comment (tropuq) [inline src/fs/block_allocator.hh]: Maybe instead of unordered_map<size_t, bool> use unordered_set<size_t> and check if elements are in that set or not?

**Review by tropuq — CHANGES_REQUESTED** [review]

File: src/fs/block_allocator.hh (hunk 2 above)
Comment (wmitros) [reply src/fs/block_allocator.hh]: good idea

File: src/fs/block_allocator.hh (hunk 1 above)
Comment (wmitros) [reply src/fs/block_allocator.hh]: for testing, will change

File: src/fs/block_allocator.hh (hunk 3 above)
Comment (wmitros) [reply src/fs/block_allocator.hh]: that would be in fact simpler

**Comment by psarna** [comment]
As for the map vs set discussion, `set` is definitely fine for now, but since blocks have fixed size, later we should even consider a bitset - with a 1TB disk and 32MB-sized chunks, bitset's size would be around 4KB and it will definitely be faster than a set. But, it's premature optimization for now, set is OK.

Also, general request - please squash all the fixup commits with `git rebase -i` and force-repush so that git history remains clear. Series for Seastar are only merged if they are clear, i.e. they don't include fixes to themselves. In this particular case, all these commits can be squashed into a single `Add block allocator` commit. Finally, please add some unittests for the feature.

File: src/fs/block_allocator.cc [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ */
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+#include "fs/block_allocator.hh"
+#include <cassert>
+
+namespace seastar {
+
+namespace fs{
```
Comment (tropuq) [inline src/fs/block_allocator.cc]:
```suggestion
namespace fs {
```

File: src/fs/block_allocator.cc [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+block_allocator::block_allocator(std::queue<size_t> fb)
+        : _allocated_blocks(), _free_blocks(std::move(fb)) {};
```
Comment (tropuq) [inline src/fs/block_allocator.cc]:
```suggestion
        : _free_blocks(std::move(fb)) {}
```

File: src/fs/block_allocator.hh [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+namespace seastar {
+
+namespace fs{
+
+class block_allocator {
+    std::unordered_set<size_t> _allocated_blocks;
+    std::queue<size_t> _free_blocks;
+public:
+    block_allocator(std::queue<size_t>);
+    ~block_allocator() = default;
```
Comment (tropuq) [inline src/fs/block_allocator.hh]: not needed i think

**Review by tropuq — CHANGES_REQUESTED** [review]

File: tests/unit/block_allocator_test.cc [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * under the License.
+ */
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+#define BOOST_TEST_MODULE fs
+
+#include <boost/test/included/unit_test.hpp>
+#include "../../src/fs/block_allocator.hh"
```
Comment (rokinsky) [inline tests/unit/block_allocator_test.cc]:
```suggestion
```

File: tests/unit/block_allocator_test.cc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+#include <deque>
+#include <seastar/core/units.hh>
+
+using namespace seastar;
+
+
+BOOST_AUTO_TEST_CASE(block_allocator) {
+    const uint64_t block = 32 * MB;
```
Comment (rokinsky) [inline tests/unit/block_allocator_test.cc]:
```suggestion
    constexpr uint64_t block = 32 * MB;
```

File: tests/unit/block_allocator_test.cc [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    small_ba.free(deq[3]);
+    BOOST_REQUIRE_EQUAL(small_ba.alloc(), deq[2]);
+    BOOST_REQUIRE_EQUAL(small_ba.alloc(), deq[4]);
+    small_ba.free(deq[2]);
+    BOOST_REQUIRE_EQUAL(small_ba.alloc(), deq[3]);
+    small_ba.free(deq[4]);
+    BOOST_REQUIRE_EQUAL(small_ba.alloc(), deq[2]);
+
+    std::queue<uint64_t> q;
+    for (uint64_t i = 0; i < 1024; i++) {
```
Comment (rokinsky) [inline tests/unit/block_allocator_test.cc]: assign `1024` to some `constexpr` value at the top of the test case.

**Review by rokinsky — CHANGES_REQUESTED** [review]

File: tests/unit/block_allocator_test.cc [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+#define BOOST_TEST_MODULE fs
+
+#include <boost/test/included/unit_test.hpp>
+#include "../../src/fs/block_allocator.hh"
+#include <deque>
+#include <seastar/core/units.hh>
+
```
Comment (rokinsky) [inline tests/unit/block_allocator_test.cc:28]:
```suggestion

#include "../../src/fs/block_allocator.hh"

```

**Review by rokinsky — CHANGES_REQUESTED** [review]

**Comment by psarna** [comment]
plz rebase on top of current head of `zpp_fs` and I'll merge it

File: src/fs/block_allocator.hh [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace seastar {
+
+namespace fs {
+
+class block_allocator {
+    std::unordered_set<uint64_t> _allocated_blocks;
+    std::queue<uint64_t> _free_blocks;
+public:
+    block_allocator(std::queue<uint64_t>);
```
Comment (tropuq) [inline src/fs/block_allocator.hh]: add explicit

File: tests/unit/block_allocator_test.cc [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+using namespace seastar;
+
+
+BOOST_AUTO_TEST_CASE(block_allocator) {
+    constexpr uint64_t block = 32 * MB;
+    constexpr uint64_t blocks_per_shard = 1024;
+    std::queue<uint64_t> empty_queue;
+    fs::block_allocator empty_ba{empty_queue};
+    BOOST_REQUIRE_EQUAL(empty_ba.alloc(), 0);
+    std::deque<uint64_t> deq{1*block,5*block,3*block,4*block,2*block};
```
Comment (tropuq) [inline tests/unit/block_allocator_test.cc]:
```suggestion
    std::deque<uint64_t> deq{1 * block, 5 * block, 3 * block, 4 * block, 2 * block};
```

File: tests/unit/block_allocator_test.cc [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    BOOST_REQUIRE_EQUAL(small_ba.alloc(), deq[2]);
+    BOOST_REQUIRE_EQUAL(small_ba.alloc(), deq[4]);
+    small_ba.free(deq[2]);
+    BOOST_REQUIRE_EQUAL(small_ba.alloc(), deq[3]);
+    small_ba.free(deq[4]);
+    BOOST_REQUIRE_EQUAL(small_ba.alloc(), deq[2]);
+
+    std::queue<uint64_t> q;
+    for (uint64_t i = 0; i < blocks_per_shard; i++) {
+        q.push(i*block);
```
Comment (tropuq) [inline tests/unit/block_allocator_test.cc]:
```suggestion
        q.push(i * block);
```

File: tests/unit/block_allocator_test.cc [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    fs::block_allocator ordinary_ba(q);
+    for (uint64_t i = 0; i < blocks_per_shard; i++) {
+        BOOST_REQUIRE_EQUAL(ordinary_ba.alloc(), i*block);
```
Comment (tropuq) [inline tests/unit/block_allocator_test.cc]:
```suggestion
        BOOST_REQUIRE_EQUAL(ordinary_ba.alloc(), i * block);
```

File: tests/unit/block_allocator_test.cc [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    for (uint64_t i = 0; i < blocks_per_shard; i++) {
+        ordinary_ba.free(i*block);
```
Comment (tropuq) [inline tests/unit/block_allocator_test.cc]:
```suggestion
        ordinary_ba.free(i * block);
```

**Review by tropuq — CHANGES_REQUESTED** [review]

File: tests/unit/block_allocator_test.cc [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+#define BOOST_TEST_MODULE fs
+
+#include <boost/test/included/unit_test.hpp>
+#include <deque>
+#include <seastar/core/units.hh>
+
+#include "../../src/fs/block_allocator.hh"
```
Comment (psarna) [inline tests/unit/block_allocator_test.cc]: Please use canonical include paths, not relative ones. `"fs/block_allocator.hh"` or, even better, `seastar/fs/block_allocator.hh`.

Aside from that - looks good, so after this fix I can merge it.

**Review by psarna — CHANGES_REQUESTED** [review]

**Pull request merged**